client/source.go

```diff
@@ -0,0 +1,55 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package client
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NewControllerRuntimeSource returns a reconciler for DynamicWatcher that sends events to a controller-runtime
+// source.Channel. This source.Channel can be used in the controller-runtime builder.Builder.Watches method. This
+// source.Channel will only send event.GenericEvent typed events, so any handlers specified in the
+// builder.Builder.Watches method will need to handle that.
+func NewControllerRuntimeSource(bufferSize int) (*ControllerRuntimeSourceReconciler, *source.Channel) {
+	eventChan := make(chan event.GenericEvent, bufferSize)
+	sourceChan := &source.Channel{
+		Source:         eventChan,
+		DestBufferSize: bufferSize,
+	}
+
+	return &ControllerRuntimeSourceReconciler{eventChan}, sourceChan
+}
+
+// ControllerRuntimeSourceReconciler is a reconciler to integrate with controller-runtime. See
+// NewControllerRuntimeSource.
+type ControllerRuntimeSourceReconciler struct {
+	eventChan chan<- event.GenericEvent
+}
+
+// Reconcile will convert the input ObjectIdentifier and send a controller-runtime GenericEvent on
+// ControllerRuntimeSourceReconciler's eventChan channel.
+func (t *ControllerRuntimeSourceReconciler) Reconcile(
+	ctx context.Context, watcher ObjectIdentifier,
+) (
+	reconcile.Result, error,
+) {
+	watcherObj := &unstructured.Unstructured{
```

Comment:
Using `unstructured` here seems easier code-wise, but I wonder if it possibly has any performance impacts over a very slim struct implementing Object that we define ourselves. Not something to change now, just want to note a difference between what you have here and what I had in my prototype